test/controllers/authentications/microsoft_graph_oauth_controller_test.rb

```diff
@@ -0,0 +1,182 @@
+require "test_helper"
+
+class Authentications::MicrosoftGraphOauthControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    OmniAuth.config.test_mode = true
+  end
+
+  test "cancelling a microsoft oauth flow redirects to the settings page" do
+    login_as users(:keith)
+    get "/auth/failure"
+    assert_redirected_to edit_settings_person_path
+  end
+
+  test "cancelling a microsoft oauth flow redirects to the login" do
+    get "/auth/failure"
+    assert_redirected_to login_path
+  end
+
+  test "should add a MicrosoftGraphCredential and redirect to edit_settings" do
+    # rob has no microsoft credential
+    user = users(:rob)
+    login_as user
+    details = {
+      uid: "rob-abc123",
+      credentials: {
+        token: "abc123",
+        refresh_token: "xyz789",
+        scope: "openid profile email offline_access user.read mailboxsettings.read files.read.all"
+      },
+      info: { email: "rob.personal@gmail.com" }
+    }
+    OmniAuth.config.add_mock(:microsoft_graph, details)
+
+    refute user.microsoft_graph_credential.present?
+    get microsoft_graph_oauth_path
+
+    assert_redirected_to edit_settings_person_path
+    assert_equal "Saved", flash[:notice]
+    assert_credential_matches_details(user.reload.microsoft_graph_credential, details)
+  end
```

Comment:
This is the same `if Current.user` settings-page case that we're not supporting so we should remove this test.

Reply:
As above, I believe this is a useful flow for an existing User adding a connection/credentials to Microsoft.